README
(c) 1997, 1998, 1999, 2004, 2005, 2006, 2008, 2009, 2010, 2014
  Fred Lindberg, lindberg@id.wustl.edu (code+docs+faq)
  Fred B. Ringel, fredr@rivertown.net (faq)
  Bruce Guenter, bruce@untroubled.org (code+docs)

ezmlm-idx is a mailing list manager providing fully automated bounce
handling, message archiving, multi-message threaded archive retrieval,
digests, remote administration, message and subscription moderation,
subscriber-only restrictions, message trailer, subject prefix,
customizable setup, multi-language and MIME support, and more. It
includes optional support for MySQL, PostgreSQL, and SQLite subscriber
databases in addition to the standard flat-file database.

This software is distributed under the GNU General Public Licence as
published by the Free Software Foundation.  See the file LICENSE for
the conditions under which this software is made available.

Like any other piece of software (and information generally), ezmlm-idx
comes with NO WARRANTY.

This software originated from ezmlm-0.53, written by Dan J. Bernstein,
which has now been placed into the public domain. ezmlm provided basic
message distribution, subscriber address handling, message archiving
(single message retrieval only), and bounce handling.

ezmlm-idx adds multi-message threaded archive retrieval, digests, remote
administration, message and subscription moderation, subscriber-only
restrictions, message trailer, subject prefix, subscriber name storage,
customizable setup, multi-language and MIME support and more. In
addition, improved bounce handling, case-insensitive subscriber address
storage, support for distributed lists (main + sublist transparent to
the subscriber), and alternate subscriber databases.

See INSTALL for installation instructions.
See UPGRADE to upgrade from a previous version of ezmlm-idx.
See CHANGES to see what's new in this version.
See FAQ for more info on ezmlm(-idx) functions and setup.
See FILES for a list of files in this package.

NOTE: Some ezmlmrc translations are not up to date.  Please check the
status of the translations at http://ezmlm.untroubled.org/lang/ and if
your language is out of date, consider submitting corrections.  Your
contribution will be acknowledged below in the next ezmlm-idx release.

The RPM will put ezmlm-cgi in the bin directory. To use it, you must manually
move it to a cgi-bin directory, set ownership and SUID (if needed), and
set up /ezmlm/ezcgirc.

See http://cr.yp.to/ezmlm.html for the latest information about ezmlm.

See http://ezmlm.untroubled.org/ for the latest version of ezmlm-idx, as
well as on-line versions of the FAQ and docs.

Mail ``ezmlm-subscribe@lists.untroubled.org'' to join the ezmlm mailing
list (low traffic). This list is run by Bruce Guenter using ezmlm-idx.
The archive is viewable at http://lists.untroubled.org/?list=ezmlm

Send general comments, questions, bug reports, and patches to:
	ezmlm@lists.untroubled.org

ACKNOWLEDGEMENTS

Translations have been provided by:

TRANSLATIONS (from newest to oldest)
nl:	Arthur Muller <Arthur@FutureQuest.net>
it:	Tullio Andreatta <t.andreatta@troppoavanti.it>
nl:	Maurice Snellen <maurice@hypercube.demon.nl>
de:	Lars Kruse <lists@sumpfralle.de>
id:	PakOgah <pakogah@pala.bo-tak.info>
es:	Ruben Cardenal <ruben@ruben.cn>
sv:	Mads E. Eilertsen <mads.e.eilertsen@hist.no>
fr:	Jean-Marc Lasgouttes

Older translation credits:
de:	Arthur Muller <Arthur@FutureQuest.net>
        Fei Weberhofer <feisaskia@weberhofer.at>
	Johannes Weberhofer <jweberhofer@weberhofer.at>
        Frank Tegtmeyer
ja:	Masashi Fujita (objectx@polyphony.scei.co.jp)
da:	Torben Fjerdingstad (unitfj@tfj.uni-c.dk)
fr:	Frank Denis (j@industrie.capgemini.fr)
sv:	Sebastian Andersson (sa@hogia.net)
pl:	Sergiusz Pawlowicz (ser@arch.pwr.wroc.pl)
pt_BR:	Wanderlei Antonio Cavassin (cavassin@connectiva.com.br)
cs:	Jan Kasprzak (kas@informatics.muni.cz)
ru:	Roman V Isaev (rm@techno.ru)
id:	Aria Prima Novianto (aria@isnet.org)
it:	Roberto De Carlo (rodeca@flashnet.it)
        Andrea Pescetti <pescetti@apache.org>
es:	Vicent Mas, Francesc Alted, Sonia Lorente, and Cyndy DePoy
cn_GB:	HaiFeng Guo (haifeng@ms.lawton.com.cn)
hu:	Julian Severn-nek (http://js.hu/)
nl:	Willem <willem@king-pin.nl>
        Maurice Snellen <maurice@hypercube.demon.nl>


CHANGES
- Switched to standard C library functions for byte_* and str_*.

- Fixed tests to work on shells with neither echo -n nor echo \c.

- Moved most source files into subdirectories.

- Fixed compile failure on Illumos.
  Thanks Amitai Schleier

ezmlm-idx-7.2.2, 2014-05-14
===========================

- Added replytolist feature to ezmlm-send. When enabled, strips incoming
  Reply-To: header and adds its own. Also alters the behavior of
  rewritefrom.

- Fixed off-by-one typo in ezmlm-weed causing an "out of memory" error.

ezmlm-idx-7.2.1, 2014-05-09
===========================

- Fixed header address extraction in the presence of double quotes.

- Fixed rewritefrom feature to run if the list is not indexed.

ezmlm-idx-7.2.0, 2014-05-02
===========================

- Added <#C#>, <#T#>, and <#X#> substitutions for (un)subscribe
  confirmation emails, replaced with the hash code, time stamp, and
  action respectively.

- Added option to ezmlm-[un]sub to use a tag other than "manual" in Log.

- Added option to ezmlm-manage to show what response is being sent.

- Added optional rewritefrom feature to ezmlm-send, automatically
  enabled when the sender has a "reject" DMARC policy.

- Fixed behavior of ezmlm-manage -Q flag to match man page.

- Fixed ezmlm-import failing to flush output to the last message,
  and enhanced it to allow reading the mbox from stdin.
  Thanks to Tullio Andreatta.

- Fixed ezmlm-manage notifying target of an unsubscribe by a remote
  administrator when the -N option is in use.
  Thanks to Nebojsa Milovanovic.

- Fixed ezmlm-archive corrupting output index files.

- Internal rewrite of SQL modules to merge all common code.

ezmlm-idx-7.1.1, 2010-11-18
===========================

- Fixed bug in getln2 function that prevented error handling.

- Fixed bug in ezmlm-cron that prevented parsing of spaces.

- Fixed unclosed file descriptor on error path in ezmlm-cgi.

ezmlm-idx-7.1.0, 2010-11-03
===========================

- Added support for SQLite3 subscriber databases.
  Thanks to Mike Tedder.

- Added support for wildcard addresses in all address databases.

- Added support for qmqpservers to all ezmlm-idx programs.

- Added support for decoding sender addresses mangled with a BATV "btv1" tag.

- Added support for custom subject lines to ezmlm-get (for digests).

- Added a new "omitbottom" control file, equivalent to the -B option for
  both ezmlm-get and ezmlm-manage.

- Modified the (un)subscribe procedure to not require confirmation
  before telling the sender they were already on (or off) the list.

- Renamed to "setup" makefile target to the more standard "install".

- Modified ezmlm-split to exit silently if there is not "split" file,
  and added it back into the "manager" file to properly handle subscribe
  and unsubscribe requests destined for sublists.

- Made the ezmlm-reject -h (obey headerreject) option the default.

- Fixed ezmlm-send to obey the "addtrailer" setting instead of looking
  for a "text/trailer" file.

- Fixed handling of adding the trailer on old lists.

- Fixed adding the trailer to posts encoded with base64.

- Fixed handling Received: headers with the date stamp on a separate line.

- Fixed bug in command-line option processing which caused sender-
  confirmed posting to fail.

- Fixed handling BATV sender addresses with upper-case hex tags.
  Thanks to Kyle Wheeler for pointing this out.

- Adjusted filename in digest attachments to fix problem with Outlook.
  Thanks to Glen Stewart

ezmlm-idx-7.0.2, 2009-08-09
===========================

- Fixed selection of [un]subscribe subject line for moderated
  subscriptions.

- Clarified ezmlm-reject man page to note that Precedence: bulk is also
  rejected.

- Added a proper charset for ch_GB (Chinese GuoBiao), and added an
  explicit charset for the other languages that were missing them.

- Switched all ISO-8859-1 charsets to the more modern ISO-8859-15.

- Many small text updates and tweaks.

ezmlm-idx-7.0.1, 2009-07-09
===========================

- Added support for decoding sender addresses mangled by BATV prvs.

- Added missing ezmlm-checksub to the installed programs.

- Improved the TXT_BY message used in digests to work better with
  non-English languages.

- Fixed out-of-memory resulting in a permanent error.

- Fixed a buffer underflow bug in concatHDR.

- Fixed a bug in ezmlm-issubn.c that was causing it to scan the main
  list when it shouldn't.

- Fixed missing defaults in ezmlm-tstdig when digsize, digcount, or
  digtime were not created.

- Clarified the steps necessary when upgrading in UPGRADE.

ezmlm-idx-7.0.0, 2008-06-16
===========================

This version has three major architectural changes:

1. There is now support for internationalized messages.  All error
messages, all subject lines, and a few other messages are now configured
in a file named "text/messages" that is read at run time.  Messages in
this file will undergo header-style substitution before they are output.

Unlike other text files, all 3 of the files (DIR/text/messages,
/etc/ezmlm/LANG/text/messages, and /etc/ezmlm/default/text/messages) are
read, and only the first match (in the above order) is used for any
given message.  This allows for creation of partial files to override
just select messages.  If no match is found, the internal English text
is used.

2. The use of the "flags" file has been deprecated completely by
individual flag files accessed by the appropriate programs.  This, along
with corresponding changes to the ezmlmrc template, allows for lists to
be reconfigured without invoking ezmlm-make.  To upgrade a list to the
new setup, simply run "ezmlm-make -+ DIR".  Note that this will delete
the "config" and "flags" files, as they have been superceded by other
files.

3. Creation of subscriber tables has been moved out of ezmlm-mktab-* and
into ezmlm-make by adding some additional hooks to the subdb plugins.  A
ezmlm-rmtab program is also added to remove subscriber tables, also with

idx.h
#ifndef IDX_H
#define IDX_H

/* Range for '-thread' to protect large archives. A '-thread' search */
/* will start at most THREAD_BEFORE messages before the action argument */
/* and go to at most THREAD_AFTER messages after the action argument. */
#define THREAD_BEFORE 2000
#define THREAD_AFTER 2000

/* Maximum number of messages returned by get */
/* You also have to update /text/bottom in ezmlmrc if you change this */
#define MAXGET 100

/* Number of messages before latest digest to return for list-get.99999_x */
/* This is still subject to the MAXGET restriction */
#define HISTGET 30

/* Maximum subject index entries returned by index */
/* Must be multiple of 100 */
/* You also have to update /text/bottom in ezmlmrc if you change this */
#define MAXINDEX 2000

/* Max dir/text file size allowed by -edit */
#define MAXEDIT 10240

/* Timeout in seconds before a bounce warning is sent. Default is 10 */
/* days. Setting it lower reduces the number of messages in the bouce */
/* dir, but makes it more likely that an address is unsubscribed due to */
/* a temporary error. This compile-time default should rarely need */
/* changing, as it can be overridden with the ezmlm-warn -t switch */
#define BOUNCE_TIMEOUT 10L

/* ezmlm-limit defaults. Convert to moderation or defer if more than  */
/* LIMMSG messages arrive within LIMSECS */
#define LIMMSG 30L
#define LIMSECS 3600L

/* Command names and alternative command names */
/* all alternates must be defined! */
/* The language-specific blocks need to undefine and redefine commands */

#define ALT_LIST "list"
#define ALT_LISTN "listn"
#define ALT_EDIT "edit"
#define ALT_FAQ "faq"
#define ALT_GET "get"
#define ALT_HELP "help"
#define ALT_INDEX "index"
#define ALT_INFO "info"
#define ALT_LOG "log"
#define ALT_REQUEST "request"
#define ALT_SUBSCRIBE "subscribe"
#define ALT_THREAD "thread"
#define ALT_UNSUBSCRIBE "unsubscribe"
#define ALT_QUERY "query"

/* to get alternative command names, you need to undefine and redefine */
/* them. Do this within a language block and send it to the author for */
/* inclusing in future versions. If it's here already, just uncomment the */
/* define for the appropriate language. */
/* #define LANG_FR 1 */

/* French Version */
#ifdef LANG_FR
#undef ALT_SUBSCRIBE
#define ALT_SUBSCRIBE "-inscription"
#undef ALT_UNSUBSCRIBE
#define ALT_UNSUBSCRIBE "-desinscription"
#endif
/* end French Version */

/* Text that is used in the outgoing messages (there is some other text, but */
/* it needs to stay constant in order to comply with rfc1153 */

/* Since this is now run-time configurable, we'll go with the lowest */
/* common denominator (per rfc2046). -> ISO-8859-1 if you don't like that */
#define TXT_DEF_CHARSET "us-ascii"

/* Output formats - letter used to override default */
#define FORMATS "mrvnx"
#define MIME 'm'
#define RFC1153 'r'
/* ---------------- virgin = MIME without header processing */
#define VIRGIN 'v'
/* NATIVE 'n' = VIRGIN without threading */
#define NATIVE 'n'
/* MIXED => multipart/mixed MIME instead of multipart/digest. Needed to bypass*/
/* pine bug when content-transfer-encoding is used (pine fails to show the */
/* initial encoded text/plain part of mulpart/digest, but not of ../mixed) */
#define MIXED 'x'
/* default output format. */
#define DEFAULT_FORMAT MIME

/* Use MIME enclosure for message to moderate by default (1) or not (0) */
/* ezmlm-store switches -m/-M override */
#define MOD_MIME 1

/* Used to add "filname=listname.msgno" to digest part content-type line.
   This confuses the heck out of Outlook Express 5.0. To circumvent this
   bug the addition has been removed. Uncomment the next line to get it
   anyway. */
/* #define DIGEST_PART_FILENAME */

/* Mode of messages in archive. For ezmlm-0.53 this is 0744, but for  */
/* "secret" lists it may make more sense to make it 0700.             */
#define MODE_ARCHIVE 0744

/* ezmlm-get actions  (ACTION_GET also for -get in ezmlm-manage) */
#define ACTION_GET "get"
#define ACTION_INDEX "index"
#define ACTION_THREAD "thread"

/* ezmlm-request actions */
#define ACTION_REQUEST "request"

/* actions for post acceptance/rejection */
#define ACTION_ACCEPT "accept-"
#define ACTION_REJECT "reject-"

/* actions for post confirmation/discard */
#define ACTION_CONFIRM "confirm-"
#define ACTION_DISCARD "discard-"
 
/* ezmlm-manage actions */
#define ACTION_LIST "list"
#define ACTION_LISTN "listn"
#define ACTION_HELP "help"
#define ACTION_INFO "info"
#define ACTION_FAQ "faq"
#define ACTION_LOG "log"
#define ACTION_SUBSCRIBE "subscribe"
#define ACTION_UNSUBSCRIBE "unsubscribe"
#define ACTION_QUERY "query"
#define ACTION_EDIT "edit"
/* if you change this, you MUST ADJUST LENGTH_ED as well! */
#define ACTION_ED "ed."
#define LENGTH_ED 3

/* ACTION_XC has to be a string "-xc." where x is any letter. All commands */
/* should have different letters. They no longer have to match the first    */
/* letter of subscribe/unsubscribe. */
/* The third char of ACTION_SC/TC/UV/VC has to be 'c' */

/* user subscription confirm */
#define ACTION_SC "sc."
/* moderator subscription confirm */
#define ACTION_TC "tc."
/* moderator initiated subscription confirm */
#define ACTION_RC "rc."
/* user unsubscribe confirm */
#define ACTION_UC "uc."
/* moderator unsubscribe confirm */
#define ACTION_VC "vc."
/* moderator initiated unsubscribe confirm */
#define ACTION_WC "wc."

/* name addition for digest, i.e. list-"digest" Don't change! */
#define ACTION_DIGEST "digest"

/* name addition for dir/extra db, i.e. list-"allow" */
#define ACTION_ALLOW "allow"
/* name addition for dir/blacklist db, i.e. list-"deny" */
#define ACTION_DENY "deny"

/* defaults for message time out in moderation queue. If modsub is 0 */
/* or empty, DELAY_DEFAULT is used. If it is set, it is made to be  */
/* within DELAY_MIN .. DELAY_MAX. All in hours. */
#define DELAY_MIN 24
#define DELAY_DEFAULT 120
#define DELAY_MAX 240

/* Mode of messages in moderation queue. The owner mode is |'d with 7.*/
/* The group/world mode can be set to anything, but it really doesn't */
/* make sense to make these messages visible to anyone else.          */
#define MODE_MOD_MSG 0700

/* name of ezmlmrc file, added to directory names */
#define TXT_EZMLMRC "/ezmlmrc"

/* name of default configuration directory, added to auto_etc */
#define TXT_DEFAULT "/default"

/* same in dot dir for local config (-c) */
#define TXT_DOTEZMLMRC ".ezmlmrc"

/* name of config file for ezmlm-cron */
#define TXT_EZCRONRC "ezcronrc"

/* default timestamp for ezmlm-limit */
#define TXT_LOOPNUM "loopnum"

/* ezmlm-cgi config file, added to auto_etc, for normal SUID root install */
#define EZ_CGIRC "/ezcgirc"

/* ezmlm-cgi config file for local install we expect to find the file in PWD */
#define EZ_CGIRC_LOC ".ezcgirc"

/* default charset for ezmlm-cgi [config file overrides per list] */
#define EZ_CHARSET "iso-8859-1"

/*------------ Specific to SQL version ------------------------------*/
/* cookie tag for SQL version of sublisting */
/* NOTE: Need to include terminal space! */
#define TXT_TAG "X-Ezauth: "

/* max no of bounces that ezmlm-receipt stores */
#define MAX_MAIN_BOUNCES 50

/* Length of domain field for SQL version. It does only the text after */
/* the last '.' in the address, so there is no reason to set it to */
/* anything other than '3'. We truncate it rather than relying on the */
/* SQL Server since we can't be sure that the SQL Server doesn't have */
/* buffer overrun holes and the address is user-controlled */
#define DOMAIN_LENGTH 3

/* programs used for outgoing mail. Normally, qmail-queue is used. Replace */
/* with qmail-qmqpc to use only qmqp for outgoing mail. QMQPC is for */
/* large lists when DIR/qmqpservers is present. Only posts and digests will */
/* use QMQP. If the normal qmail-qmqpc is used the contents of */
/* DIR/qmqpcservers are ignored. With a patch, qmail-qmqpc will use the */
/* servers on it's command line. In this case, the IP addresses listed one */
/* per line in DIR/qmqpservers will be tried until a working one is found. */
/* the option is mainly to allow large list clusters on a single host to use */
/* different QMQPC hosts as exploders.*/
#define PROG_QMAIL_QUEUE "bin/qmail-queue"
#define PROG_QMAIL_QMQPC "bin/qmail-qmqpc"

/*---------- Things below this line are not configurable -----------*/
/* file in DIR that has the qmqpc servers (if any) */
#define QMQPSERVERS "qmqpservers"
/* database types */
#define FLD_DIGEST 1
#define FLD_ALLOW 2
#define FLD_DENY 3
/* Action types */
#define AC_NONE 0
#define AC_GET 1
#define AC_DIGEST 2
#define AC_THREAD 3
#define AC_INDEX 4
#define AC_LIST 5
#define AC_HELP 6
#define AC_EDIT 7
#define AC_DENY 8
#define AC_LOG 9
#define AC_SUBSCRIBE 10
#define AC_UNSUBSCRIBE 11
#define AC_SC 12
#define AC_LISTN 13

typedef struct msgentry {	/* one per message in range */
  unsigned long subnum;		/* subject number */
  unsigned long authnum;	/* message author number */
  unsigned int date;		/* yyyymm as number */
} msgentry;

typedef struct subentry {	/* one per unique subject in message range */
  void *higher;
  void *lower;
  char *sub;			/* string with terminating '\0' */
				/* when building, higher/lower=0 marks end */
				/* of branch. When printing, start at the  */
				/* beginning of the table and go up until  */
				/* sub = 0. */
  unsigned int sublen;
  unsigned long firstmsg;	/* the first message with this subject*/
  unsigned long lastmsg;	/* the last message with this subject*/
  unsigned char msginthread;	/* number of messages seen in this thread */
} subentry;

typedef struct authentry {	/* one per unique author in message range */
  void *higher;
  void *lower;
  char *auth;			/* string with terminating '\0' */
				/* when building, higher/lower=0 marks end */
				/* of branch. When printing, start at the  */
				/* beginning of the table and go up until  */
				/* auth = 0. */
  unsigned long authlen;
  unsigned long firstmsg;	/* the first message with this author */
				/* lastmsg not very useful as author are less */
				/* clustered than threads */
} authentry;

typedef struct dateentry {	/* date yyyymm and 1st message of that date */
  unsigned int date;
  unsigned int msg;
} dateentry;

#endif
